### PR TITLE
Runtime properties

### PR DESCRIPTION
**Pull request by iocanel**

### CODE REVIEW

File: container/tomcat/webapp/pom.xml [hunk 1]
```diff
@@ -200,7 +200,31 @@
...
+                                OSGI-INF/io.fabric8.spi.internal.ContainerManagerImpl.xml,\
+                                OSGI-INF/io.fabric8.spi.internal.ProfileManagerImpl.xml,\
+                                OSGI-INF/io.fabric8.spi.internal.ProfileBuilderService.xml,\
+                                OSGI-INF/io.fabric8.spi.internal.ProfileVersionBuilderService.xml,\
+                                OSGI-INF/io.fabric8.spi.permit.PermitManagerImpl.xml,\
+                                OSGI-INF/org.jboss.gravia.provision.internal.ProvisionerService.xml,\
+                                OSGI-INF/org.jboss.gravia.repository.internal.RepositoryMBeanService.xml,\
+                                OSGI-INF/org.jboss.gravia.repository.internal.RepositoryService.xml,\
+                                OSGI-INF/org.jboss.gravia.resolver.internal.ResolverService.xml
+                            </Service-Component>
```
Comment (tdiesler) [inline container/tomcat/webapp/pom.xml:227]: This inserts a manifest header with whitespace - it fails when the limited manifest line boundary happens to be within the whitespace

File: container/wildfly/extension/pom.xml [hunk 2]
```diff
@@ -109,7 +109,31 @@
...
+                                OSGI-INF/io.fabric8.spi.internal.ProfileManagerImpl.xml,\
+                                OSGI-INF/io.fabric8.spi.internal.ProfileBuilderService.xml,\
+                                OSGI-INF/io.fabric8.spi.internal.ProfileVersionBuilderService.xml,\
+                                OSGI-INF/io.fabric8.spi.permit.PermitManagerImpl.xml,\
+                                OSGI-INF/org.jboss.gravia.provision.internal.ProvisionerService.xml,\
+                                OSGI-INF/org.jboss.gravia.repository.internal.RepositoryMBeanService.xml,\
+                                OSGI-INF/org.jboss.gravia.repository.internal.RepositoryService.xml,\
+                                OSGI-INF/org.jboss.gravia.resolver.internal.ResolverService.xml
+                            </Service-Component>
                         </manifestEntries>
```
Comment (tdiesler) [inline container/wildfly/extension/pom.xml:137]: This inserts a manifest header with whitespace - it fails when the limited manifest line boundary happens to be within the whitespace

File: fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ */
+
+package io.fabric8.spi;
+
+import io.fabric8.api.Identifiable;
+
+import java.util.Map;
+
+public interface RuntimeService extends Identifiable<String> {
```
Comment (tdiesler) [inline fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java]: why would this need to be identifiable?

File: fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    String getConfigDirectory();
+
+    String getDataDirectory();
+
+    String getProperty(String key);
+
+    String getProperty(String key, String defaultValue);
+
+    void setProperty(String key, String value);
```
Comment (tdiesler) [inline fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java]: Should this really be read-write? Would this survive server restart? Would this be part of the data content?

File: fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+public interface RuntimeService extends Identifiable<String> {
+
+    String ID = "runtime.id";
+    String HOME_DIR = "runtime.home";
+    String DATA_DIR = "runtime.data";
+    String CONF_DIR = "runtime.conf";
+
+    String getPath();
+
+    String getConfigDirectory();
```
Comment (tdiesler) [inline fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java]: These should probably return Path or File objects

File: fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Map;
+
+public interface RuntimeService extends Identifiable<String> {
+
+    String ID = "runtime.id";
+    String HOME_DIR = "runtime.home";
+    String DATA_DIR = "runtime.data";
+    String CONF_DIR = "runtime.conf";
+
+    String getPath();
```
Comment (tdiesler) [inline fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java]: We have the AbstractManagedCreateOptions.getTargetDirectory() which is the root directory of the server. Would this be the same? If so we would need a guarantee.

File: fabric/spi/src/main/java/io/fabric8/spi/internal/ComponentConfigurer.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.felix.scr.annotations.Deactivate;
+import org.apache.felix.scr.annotations.Reference;
+import org.apache.felix.scr.annotations.Service;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Component(immediate = true)
+@Service(Configurer.class)
+public class ComponentConfigurer extends AbstractComponent implements Configurer {
```
Comment (tdiesler) [inline fabric/spi/src/main/java/io/fabric8/spi/internal/ComponentConfigurer.java]: Does this really need to be an extension point to the system? I'd suggest we keep this an implementation detail until there is a call for pluggability.

File: fabric/spi/src/main/java/io/fabric8/spi/utils/IllegalStateAssertion.java [hunk 8]
```diff
@@ -36,6 +36,15 @@ private IllegalStateAssertion() {
     /**
      * Throws an IllegalStateException when the given value is not null.
      */
+    public <T> T checkNotNull(T value, String message) {
```
Comment (tdiesler) [inline fabric/spi/src/main/java/io/fabric8/spi/utils/IllegalStateAssertion.java]: duplicates assertNotNull (with return value). If its decided to provide return values for assertions it should be done consistently for all assertions.

File: fabric/spi/src/main/java/io/fabric8/spi/utils/PlaceholderUtils.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
+/*
```
Comment (tdiesler) [inline fabric/spi/src/main/java/io/fabric8/spi/utils/PlaceholderUtils.java]: Generally, the more stuff we can make implementation private the less we have to worry about compatibility. Are all of these utils really meant to be part of the public API?

File: container/wildfly/extension/pom.xml (hunk 2 above)
Comment (iocanel) [inline container/wildfly/extension/pom.xml:137]: What are our alternatives? The previous version was not readable.

File: fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java (hunk 3 above)
Comment (iocanel) [inline fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java]: You are right it doesn't need to and the interface should be removed.

File: fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java (hunk 6 above)
Comment (iocanel) [inline fabric/spi/src/main/java/io/fabric8/spi/RuntimeService.java]: Yes its the same, we need to check how we can guarantee that the two are the aligned.

**Pull request closed without merging**